server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/CompactableImpl.java

```diff
@@ -214,6 +202,9 @@ public CompactableImpl(Tablet tablet, CompactionManager manager,
       }
       return Set.copyOf(servicesIds);
     }, 2, TimeUnit.SECONDS);
+
+    this.fileMgr = new TabletCompactionFileManager(Collections.unmodifiableSet(runningJobs),
+        tablet.getExtent(), extCompactingFiles, extSelInfo);
```

Comment:
This creates an unmodifiable view over `runningJobs`. However, `runningJobs` can still be modified in the below synchronized methods and this view doesn't protect against concurrent reads over the set, which could lead to `ConcurrentModificationException`s while streaming/iterating the set in the new `TabletCompactionFileManager` class. I think instead of using a `HashSet` and synchronize methods, you could use `ConcurrentHashMap.newKeySet()` for the `runningJobs` field. You can still wrap it with unmodifiable to ensure `TabletCompactionFileManager` can't alter the set.

Reply:
Currently all accesses to runningJobs and fileMgr are within in synchronized blocks in CompactableImpl (at least that is my expectation). This is because changes in state related to compactions must atomically change the jobs, files, and other information.  Therefore a ConcurrentHashMap would not be necessary and using it would imply a usage pattern that is not wanted.  I will document this expectation.

Reply:
I guess it's technically correct if all operations on `TabletCompactionFileManager` instances are guarded with a synchronized block as well. However, my concern is that's very fragile and hard to enforce. It also implies a high degree of coordination between the two separate classes in one sense (for synchronization), whereas in another sense (mutable data access), that degree of coordination is implied by the protective use of `Collections.unmodifiableSet` to not exist. If they were coordinating to that degree (such as if `TabletCompactionFileManager` were an inner-class of `CompactableImpl`), I wouldn't think it necessary to do that protective wrapping.

Additionally, it feels weird that methods on the `TabletCompactionFileManager` object would require synchronizing on `CompactableImpl.this`. This feels like `CompactableImpl`'s internal implementation details are leaking into the second object.

I just think it might be more clear if, as long as we're separating these into two distinct classes, we further ensure they are decoupled, so you don't have to make assumptions about how the set of running jobs is shared between them. As is, the concurrency model seems very fragile. Some ideas:

1. Synchronize on the `runningJobs` object instead of `CompactableImpl.this`, or via some other very explicit lock object,
2. Don't share `runningJobs` with `TabletCompactionFileManager` at all, keeping all synchronization and access to `runningJobs` internal to `CompactableImpl`; expose what you need through `CompactableImpl` API methods

With lambdas, it seems we should be able to get very explicit, and less error prone synchronization, all contained in `CompactableImpl`, without leaking anything out:

```java
public interface CompactableImpl {
  // ...
  public synchronized <T> T synchronizedDo(Function<Set<CompactionJob>, T> operation) {
    return operation.apply(runningJobs);
  }
  // ...
}
```

Reply:
The classes are meant to be tightly coupled.  It was never intended that anything besides CompactableImpl would use TabletCompactionFileManager.  Making it an inner class seems like a good way to express this.  Tracking the state of compacting files and which files are eligible for compaction is one of the most complicated parts of CompactableImpl currently.  The goal with this change was to logically group that code to make it easier to test and modify.  However I don't want to attempt to make it a standalone unit in terms of functionality or synchronization as I think doing this would make the code harder to understand and verify for correctness.  I can not really think of anything simpler to understand than the current synchronized code blocks in CompactableImpl.  Making it an inner class will make it clear that is tightly coupled while still allowing testing and logical grouping of the code that does bookkeeping for compacting files.

Reply:
Made the class an inner class in e035a3f.  In this change it no longer passes in runningJobs, the inner class just calls a method on the outer class to access it.  Also made the name of the inner class FileManager

Reply:
I think the inner class makes sense. However, it's still hard to verify that all the accesses are synchronized. For example, `noneRunning` has a read on `runningJobs` but is not synchronized. Tracing it back, it is called by some other methods which are also not synchronized... but then those methods are called by other methods which are called only in synchronized blocks. Requiring all these uses to be traced back to verify the concurrency model is correct is frustrating and error-prone. I wish there was a better way, but not sure of what could work better. Maybe you could throw in an extra synchronized keyword or two to guard `runningJobs`, so it's less error-prone, taking advantage of reentrancy so the redundant synchronized keywords don't matter. Either way, I think the tight coupling is more clear as an inner-class. Thanks for making that change.

Reply:
> I wish there was a better way, but not sure of what could work better.

@ctubbsii  not sure if this workable, but one possible way is to move all in memory state in compactableImpl into another class that handles all sync.  A lot of the code in compactableImpl has the following pattern.

```java
someMethod() {
   synchronized(this) {
       // examine/change multiple instance var that change the objects state
   }

   // do something like a metadata read/write or call to Tablet
  
  synchronized(this) {
       // examine/change multiple instance var that change the objects state
  }
```

The above sync blocks could be replaced w/ calls to an object that synchronized and manages all of the in memory state for CompactableImpl.  The changes in this PR take a subset of that in memory state (files) and move them to an inner class.  

Would need to analyze the code to see if its possible, but it could be one way to simplify the code by grouping the in memory state and sync into a single class (excluding the I/O type operations).